apollo-router-core/src/layer.rs

```diff
@@ -0,0 +1,101 @@
+use crate::{SubgraphRequest, SubgraphResponse};
+use async_trait::async_trait;
+use once_cell::sync::Lazy;
+use schemars::gen::SchemaGenerator;
+use schemars::JsonSchema;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex, MutexGuard};
+use tower::util::{BoxLayer, BoxService};
+use tower::BoxError;
+
+#[derive(Clone)]
+pub struct LayerFactory {
+    instance_factory: InstanceFactory,
+    schema_factory: SchemaFactory,
+}
+
+type BoxedSubgraphLayer = BoxLayer<
+    BoxService<SubgraphRequest, SubgraphResponse, BoxError>,
+    SubgraphRequest,
+    SubgraphResponse,
+    BoxError,
+>;
+
+impl LayerFactory {
+    pub fn new(instance_factory: InstanceFactory, schema_factory: SchemaFactory) -> Self {
+        Self {
+            instance_factory,
+            schema_factory,
+        }
+    }
+
+    pub fn create_instance(
+        &self,
+        configuration: &serde_json::Value,
+    ) -> Result<BoxedSubgraphLayer, BoxError> {
+        (self.instance_factory)(configuration)
+    }
+
+    pub fn create_schema(&self, gen: &mut SchemaGenerator) -> schemars::schema::Schema {
+        (self.schema_factory)(gen)
+    }
+}
+
+type InstanceFactory = fn(
+    &serde_json::Value,
+) -> Result<
+    BoxLayer<
+        BoxService<SubgraphRequest, SubgraphResponse, BoxError>,
+        SubgraphRequest,
+        SubgraphResponse,
+        BoxError,
+    >,
+    BoxError,
+>;
+
+type SchemaFactory = fn(gen: &mut SchemaGenerator) -> schemars::schema::Schema;
+
+static LAYER_REGISTRY: Lazy<Mutex<HashMap<String, LayerFactory>>> = Lazy::new(|| {
+    let m = HashMap::new();
+    Mutex::new(m)
+});
+
+pub fn layers() -> Arc<HashMap<String, LayerFactory>> {
+    Arc::new(LAYER_REGISTRY.lock().expect("Lock poisoned").clone())
```

Comment:
I'm wondering why we don't put an `Arc` directly in the `LAYER_REGISTRY` to avoid cloning the entire layer_registry ?

Reply:
Then `layers_mut` could not be used to mutate the registry.
I guess we could put an `rwlock` in?

Reply:
Done, it looks much better now.